### PR TITLE
144. Binary Tree Preorder Traversal

### PR DESCRIPTION
**Pull request by ErdemT09**
Resolves: https://leetcode.com/problems/binary-tree-preorder-traversal/

## Algorithm:
This problem is pretty simple. We could either implement a DFS algorithm that is by definition a preorder traversal. 
As for the stack approach, we put the right node of a root first and then the left node, because the stack works on the principle "Last in, First out".

### CODE REVIEW

File: src/main/java/algorithms/easy/binarytreepreordertraversal/BinaryTreePreorderTraversal.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+		preorder.add(root.val);
+
+		if (root.left != null) {
+			DFSpreorder(root.left);
+		}
+
+		if (root.right != null) {
+			DFSpreorder(root.right);
+		}
```
Comment (altayhunoglu) [inline src/main/java/algorithms/easy/binarytreepreordertraversal/BinaryTreePreorderTraversal.java:33]:
```suggestion
          if (root == null) return;
          preorder.add(root.val);
          DFSpreorder(root.left);
          DFSpreorder(root.right);
```

File: src/main/java/algorithms/easy/binarytreepreordertraversal/BinaryTreePreorderTraversalStack.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.Stack;
+
+import algorithms.datastructures.TreeNode;
+
+public class BinaryTreePreorderTraversalStack {
+
+	public List<Integer> preorderTraversal(TreeNode root) {
+		if (root == null) {
+			return Collections.emptyList();
+		}
```
Comment (altayhunoglu) [inline src/main/java/algorithms/easy/binarytreepreordertraversal/BinaryTreePreorderTraversalStack.java:15]:
```suggestion
 
```

File: src/main/java/algorithms/easy/binarytreepreordertraversal/BinaryTreePreorderTraversal.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+public class BinaryTreePreorderTraversal {
+
+	List<Integer> preorder = new ArrayList<>();
+
+	public List<Integer> preorderTraversal(TreeNode root) {
+
+		if (root == null) {
+			return Collections.emptyList();
+		}
```
Comment (altayhunoglu) [inline src/main/java/algorithms/easy/binarytreepreordertraversal/BinaryTreePreorderTraversal.java:17]:
```suggestion

```

File: src/main/java/algorithms/easy/binarytreepreordertraversal/BinaryTreePreorderTraversalStack.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+		while (!stack.isEmpty()) {
+			TreeNode node = stack.pop();
+			preorder.add(node.val);
+			if (node.right != null) {
+				stack.push(node.right);
+			}
+			if (node.left != null) {
+				stack.push(node.left);
+			}
```
Comment (altayhunoglu) [inline src/main/java/algorithms/easy/binarytreepreordertraversal/BinaryTreePreorderTraversalStack.java:30]:
```suggestion
            if(node == null) continue;
            preorder.add(node.val);
            stack.push(node.right);
	      stack.push(node.left);
```

Comment (ErdemT09) [reply src/main/java/algorithms/easy/binarytreepreordertraversal/BinaryTreePreorderTraversalStack.java:30]: I think adding null values to the stack will be less efficient than one extra null-check.

**Review by altayhunoglu — APPROVED** [review]
Thanks, Erdem. Nice and clear code.

**Pull request merged**